src/pgwire/src/protocol.rs

```diff
@@ -308,16 +314,28 @@ where
                 state = match state {
                     State::Ready => self.advance_ready().await?,
                     State::Drain => self.advance_drain().await?,
+                    State::TimedOut(err) => self.advance_timed_out(err).await?,
                     State::Done => return Ok(()),
-                }
+                };
+                self.adapter_client()
+                    .add_idle_in_transaction_session_timeout();
             }
         }
     }
 
     async fn advance_ready(&mut self) -> Result<State, io::Error> {
-        let message = self.conn.recv().await?;
+        // Handle timeouts first so we don't execute any statements when there's a pending timeout.
+        let message = select! {
+            biased;
+
+            Some(timeout) = self.adapter_client.as_mut().expect(Self::ADAPTER_CLIENT_INVARIANT).recv_timeout() =>
```

Comment:
> I didn't realize you could have match blocks on the left side of =. What happens if it's None? The select drops it and goes back to polling?

To be honest I just copied the format of how we use `select!` in the Coord: https://github.com/MaterializeInc/materialize/blob/cbbcec8324ffc845b9fc08018207d526203d81e8/src/adapter/src/coord.rs#L945-L997

However, I looked into this here: https://tokio.rs/tokio/tutorial/select
- When a branch doesn't match the pattern, i.e. returns `None` here, the branch is disabled for that invocation of `select!`. In this case that means that we block and wait for `self.conn.recv()` to return.
- The channel will only return `None` once they are closed, and once they are closed they will continue to only ever return `None`.

I'm not sure if the `None` get's dropped or if it's kept in the Channel queue, however the result is the same. Since once we see a single `None`, we'll only ever see `None`.

I also just pushed some comments about cancellation safety.

Reply:
> I have no idea if this matters but it seems like a possible minor cost to pay the expect check each time here. The recv_timeout() future should be correct to create once and always listen on it. But pretty sure the borrow checker would complain about that? Unclear to me if either of these are true and matter at all.

I wasn't able to follow what this comment means, I think I'm just not familiar enough with certain async concepts.